Sources/SwiftParser/Expressions.swift

```diff
@@ -525,21 +525,26 @@ extension Parser {
     flavor: ExprFlavor,
     pattern: PatternContext = .none
   ) -> RawExprSyntax {
-    // First check to see if we have the start of a regex literal `/.../`.
-    //    tryLexRegexLiteral(/*forUnappliedOperator*/ false)
-
-    // Try parse an 'if' or 'switch' as an expression. Note we do this here in
-    // parseUnaryExpression as we don't allow postfix syntax to hang off such
-    // expressions to avoid ambiguities such as postfix '.member', which can
-    // currently be parsed as a static dot member for a result builder.
-    if self.at(.keyword(.switch)) {
+    // Try parse a single value statement as an expression (e.g do/if/switch).
+    // Note we do this here in parseUnaryExpression as we don't allow postfix
+    // syntax to hang off such expressions to avoid ambiguities such as postfix
+    // '.member', which can currently be parsed as a static dot member for a
+    // result builder.
+    switch self.at(anyIn: SingleValueStatementExpression.self) {
+    case (.do, _)?:
       return RawExprSyntax(
-        parseSwitchExpression(switchHandle: .constant(.keyword(.switch)))
+        self.parseDoExpression(doHandle: .constant(.keyword(.do)))
       )
-    } else if self.at(.keyword(.if)) {
+    case (.if, _)?:
       return RawExprSyntax(
-        parseIfExpression(ifHandle: .constant(.keyword(.if)))
+        self.parseIfExpression(ifHandle: .constant(.keyword(.if)))
       )
+    case (.switch, _)?:
+      return RawExprSyntax(
+        self.parseSwitchExpression(switchHandle: .constant(.keyword(.switch)))
+      )
```

Comment:
I didn’t check but wouldn’t this be cleaner now as

```suggestion
    case (.switch, let handle)?:
      return RawExprSyntax(
        self.parseSwitchExpression(switchHandle: handle)
      )
```

Same for the other `.constant` cases here.

Not that we want to do it, but that would make the code portable to changing `self.at` to `self.canRecoverTo`.

Reply:
Hah yeah that's what I originally tried, but there didn't seem to be a nice way to go from TokenConsumptionHandle to RecoveryConsumptionHandle, looks like there's `.noRecovery` now, so I should be able to use that.

Reply:
Yeah, I think `noRecovery` is a recent-ish addition.